// @nativecn/cli/dist/commands/add.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addCommand = addCommand;
const path_1 = __importDefault(require("path"));
const chalk_1 = __importDefault(require("chalk"));
const execa_1 = __importDefault(require("execa"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const config_1 = require("../utils/config");
const package_1 = require("../utils/package");
const templates_1 = require("../utils/templates");
// Component-specific dependencies
const componentDependencies = {
    alert: {
        deps: ['react-native-vector-icons'],
        devDeps: ['@types/react-native-vector-icons'],
    },
};
function addCommand(components_1) {
    return __awaiter(this, arguments, void 0, function* (components, options = {}) {
        // Read the config
        const config = yield (0, config_1.readConfig)();
        // Use config output directory if not specified
        const targetDir = options.dir || config.components.outDir;
        console.log(chalk_1.default.bold(`Adding components to ${targetDir}...`));
        // Resolve all dependencies
        const resolvedComponents = (0, templates_1.resolveComponentDependencies)(components);
        const additionalComponents = resolvedComponents.filter(c => !components.includes(c));
        if (additionalComponents.length > 0) {
            console.log(chalk_1.default.blue('i'), `Adding dependencies: ${additionalComponents.join(', ')}`);
        }
        // Check if we need to install any npm dependencies for the components
        if (!options.skipDependencies) {
            yield installComponentDependencies(resolvedComponents);
        }
        // Process each component
        const results = yield Promise.all(resolvedComponents.map((component) => __awaiter(this, void 0, void 0, function* () {
            const result = yield (0, templates_1.copyComponentTemplate)(component, targetDir, {
                overwrite: options.overwrite,
            });
            return { component, success: result };
        })));
        // Generate index.ts file
        yield (0, templates_1.generateComponentIndex)(targetDir);
        // Print summary
        console.log();
        console.log(chalk_1.default.bold('Summary:'));
        const successful = results.filter(r => r.success);
        const failed = results.filter(r => !r.success);
        if (successful.length > 0) {
            console.log(chalk_1.default.green('✓'), `Added ${successful.length} component(s):`);
            successful.forEach(result => {
                console.log(`  • ${result.component}`);
            });
        }
        if (failed.length > 0) {
            console.log(chalk_1.default.red('✗'), `Failed to add ${failed.length} component(s):`);
            failed.forEach(result => {
                console.log(`  • ${result.component}`);
            });
        }
        // Show usage example
        if (successful.length > 0) {
            console.log();
            console.log(chalk_1.default.bold('Usage example:'));
            console.log();
            console.log(`import { ${successful
                .map(r => r.component.charAt(0).toUpperCase() + r.component.slice(1))
                .join(', ')} } from "${path_1.default.relative(process.cwd(), path_1.default.join(process.cwd(), targetDir))}";`);
            console.log();
        }
    });
}
/**
 * Install npm dependencies required by specific components
 */
function installComponentDependencies(components) {
    return __awaiter(this, void 0, void 0, function* () {
        // Collect all required dependencies
        const dependencies = [];
        const devDependencies = [];
        for (const component of components) {
            if (componentDependencies[component]) {
                // Regular dependencies
                for (const dep of componentDependencies[component].deps) {
                    if (!dependencies.includes(dep) && !(yield (0, package_1.isPackageInstalled)(dep))) {
                        dependencies.push(dep);
                    }
                }
                // Dev dependencies
                for (const dep of componentDependencies[component].devDeps) {
                    if (!devDependencies.includes(dep) && !(yield (0, package_1.isPackageInstalled)(dep))) {
                        devDependencies.push(dep);
                    }
                }
            }
        }
        // Detect package manager
        const packageManager = yield detectPackageManager();
        const installCmd = packageManager === 'yarn' ? 'add' : 'install';
        let success = true;
        // Install regular dependencies if any
        if (dependencies.length > 0) {
            console.log(chalk_1.default.bold(`Installing component dependencies: ${dependencies.join(', ')}...`));
            try {
                yield (0, execa_1.default)(packageManager, [installCmd, ...dependencies], {
                    stdio: 'inherit',
                });
                console.log(chalk_1.default.green('✓'), 'Component dependencies installed');
                if (dependencies.includes('react-native-vector-icons')) {
                    console.log(chalk_1.default.blue('i'), 'Additional setup for react-native-vector-icons may be required:');
                    console.log('   For iOS: pod install in the ios directory');
                    console.log('   For Android: Check that gradle is configured correctly');
                    console.log('   See: https://github.com/oblador/react-native-vector-icons#installation');
                }
            }
            catch (error) {
                console.error(chalk_1.default.red('✗'), 'Failed to install component dependencies:', error instanceof Error ? error.message : String(error));
                console.log(chalk_1.default.yellow('!'), 'You can install them manually:', `${packageManager} ${installCmd} ${dependencies.join(' ')}`);
                success = false;
            }
        }
        // Install dev dependencies if any
        if (devDependencies.length > 0) {
            console.log(chalk_1.default.bold(`Installing component dev dependencies: ${devDependencies.join(', ')}...`));
            try {
                const devFlag = packageManager === 'npm' ? '--save-dev' : '-D';
                yield (0, execa_1.default)(packageManager, [installCmd, devFlag, ...devDependencies], {
                    stdio: 'inherit',
                });
                console.log(chalk_1.default.green('✓'), 'Component dev dependencies installed');
            }
            catch (error) {
                console.error(chalk_1.default.red('✗'), 'Failed to install component dev dependencies:', error instanceof Error ? error.message : String(error));
                console.log(chalk_1.default.yellow('!'), 'You can install them manually:', `${packageManager} ${installCmd} ${packageManager === 'npm' ? '--save-dev' : '-D'} ${devDependencies.join(' ')}`);
                success = false;
            }
        }
        return success;
    });
}
/**
 * Detect the package manager being used
 */
function detectPackageManager() {
    return __awaiter(this, void 0, void 0, function* () {
        const cwd = process.cwd();
        if (yield fs_extra_1.default.pathExists(path_1.default.join(cwd, 'yarn.lock'))) {
            return 'yarn';
        }
        if (yield fs_extra_1.default.pathExists(path_1.default.join(cwd, 'pnpm-lock.yaml'))) {
            return 'pnpm';
        }
        return 'npm';
    });
}
